### ISSUE TITLE
[Ch 26][Notebook][Major] KoGPT2 reference generation 이 깨진 토큰으로 나오는데 해석 가이드는 "자연스러운 문장 흐름" 이라고 정반대 서술 (WikiDocs 인정 문단의 원인 설명도 오류)

### ISSUE THREAD

**Issue by yum-yeom** [yoon-gu/neuqes-101#221]
> **요약** — Ch 26 §8 의 KoGPT2 reference generation 이 Colab T4 에서 **깨진 토큰 나열**(`�`·`▁` 반복, 영어 인명 조각)로 나옵니다. 그런데 **노트북·EPUB 은 바로 아래 해석 가이드에서 "REF (KoGPT2 125M): *다양한 도메인 어휘 + 자연스러운 문장 흐름*" 이라고 정반대로 서술**합니다. WikiDocs 에만 이를 인정하는 문단이 있지만 **원인 설명이 틀렸고**, 그 문단 바로 다음에 같은 해석 가이드가 그대로 붙어 자기모순입니다.

## 무엇이 문제인가

기준: master `d6c78e1` · 독립 재실행 2026-09-04 (colab-cli, T4, 189.9초, `status: ok`) — **커밋 실행본(2026-06-21)과 증상 동일**

셀 `c39` (§8 선택 reference) 실제 출력:

```
REFERENCE KoGPT2 (125M) - generation on same Korean prompts
======================================================================
prompt: 옛날 옛날에
�����▁�ng��▁�ng��n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�n�

prompt: 작은 소녀가
�����i���i,▁▁▁▁▁▁▁

prompt: 큰 개가
����▁▁▁▁▁▁▁▁▁▁▁▁▁▁

prompt: 어느 날,
���,▁Gaurius,▁vand.▁Gould,▁Denol,▁Tamil,▁Anat�nn▁McCamel,▁Zehn,▁Guy.▁Cyb,▁Shitch,▁Pellek,▁Nami,▁Jalifa,
```

| 산출물 | 깨진 출력 | 인정 문단 | 해석 가이드 |
|---|---|---|---|
| **Notebook** (`c39`→`ma40`) | 있음 | **없음** | "다양한 도메인 어휘 + 자연스러운 문장 흐름" |
| **WikiDocs** (26-2, https://wikidocs.net/370143 ) | 있음 | 있음 (**원인 설명 오류**) | 인정 문단 **바로 아래** 같은 서술 → 자기모순 |
| **EPUB** (`ch032.xhtml`) | 있음 (`?????_?ng??…` — 변환기가 U+FFFD→`?`) | **없음** | 같은 서술 |

WikiDocs 인정 문단은 원인을 이렇게 적고 있습니다:

> 우리 GEN_KWARGS·prompt 인코딩이 KoGPT2 토크나이저와 잘 맞지 않은 탓으로

**이 설명은 사실과 다릅니다.** 같은 모델·같은 prompt·같은 `GEN_KWARGS`(`max_new_tokens=60, do_sample=True, temperature=0.8, top_k=50`)로 대조 실행하면 정상 한국어가 나옵니다:

- **로컬 (transformers 4.57.6, CPU)** — `옛날 옛날에` → `"…이게 네. 그게 막 그~ 이게 어~ 그 어~ 우리 그~ 사십년 전 같은 경우에는…"`, `작은 소녀가` → `"…톡톡 터뜨리곤 한다.\n그녀가 톡톡 터뜨리는 모습을 직접 보고 싶다면 먼저 여행 가서 보자…"`
- **Ch 27 실행본** (같은 `skt/kogpt2-base-v2`, 같은 prompt) — `"네들이 말했잖아. 그게 뭐냐면 우리가 그~ …"` 등 정상

토크나이저도 정상입니다 — `AutoTokenizer.from_pretrained("skt/kogpt2-base-v2")` 는 `GPT2TokenizerFast`(vocab 51,200)로 로드되고 `"옛날 옛날에" → ['▁옛날','▁옛날에'] → "옛날 옛날에"` 로 round-trip 합니다.

**관측된 단서** (단정하지 않고 로그 그대로) — 실행 로그를 보면 `from_pretrained` 는 `pytorch_model.bin` 에서 로드했다고 LOAD REPORT 를 찍는데, **generation 이 시작된 직후 `model.safetensors` (513MB) 를 다시 내려받습니다.** 생성된 토큰이 KoGPT2 vocab 안의 *무작위* 토큰(영어 인명·byte 조각)이라, 가중치가 실제로 적재되지 않은 본체로 생성한 것과 같은 모양입니다. 같은 패턴(`.to(device).eval()`)을 쓰는 Ch 24 의 `gpt2` reference 는 정상인데, `gpt2` 저장소는 safetensors 가 기본 파일이라 재다운로드가 일어나지 않습니다.

**캡처 — WikiDocs 26-2 (https://wikidocs.net/370143 , 빨간 박스)**

<img width="1568" height="639" alt="Ch 26 KoGPT2 reference 깨진 출력" src="https://github.com/user-attachments/assets/b1372861-8e74-41cb-9da3-95282d7928f4" />

> 빨간 박스 = 실제 실행 출력(`�`·`▁` 반복). 바로 아래 "결과 해석" 문단이 이를 인정하지만 원인을 `GEN_KWARGS`·prompt 인코딩 탓으로 돌리고, 그 다음 "해석 가이드" 는 여전히 REF 를 "자연스러운 문장 흐름" 으로 서술합니다.

## 왜 고쳐야 하나

이 셀의 존재 이유가 *"잘 학습된 큰 모델은 이렇게 자연스러운 한국어를 만든다"* 를 보여 주는 것입니다. 지금은 **정반대 증거를 싣고, 그 옆에 원래 하려던 주장을 그대로 남겨** 두었습니다. 학습자는 (a) KoGPT2 125M 이 저런 출력을 낸다고 배우거나, (b) WikiDocs 를 읽었다면 *원인이 `GEN_KWARGS`·prompt 인코딩* 이라는 틀린 진단을 배웁니다. 자기 T4 에서 돌려도 같은 깨진 출력이 나오므로 "내가 뭘 잘못했나" 로도 이어집니다.

## 제안

1. **(권장)** `c39` 의 모델 로드를 정상 동작하도록 교정 — 로드 완료 후 device 이동을 분리하거나, `dtype`/`device_map` 을 명시하거나, Ch 27 이 쓰는 로드 패턴에 맞춥니다. 교정 후 Colab T4 재실행 → `executed/` 갱신 → WikiDocs 재변환 → EPUB 재빌드.
2. 교정이 어려우면 `RUN_KOGPT2_REF = False` 를 기본값으로 두고, 해석 가이드의 REF 항목과 §8 도입부를 *Ch 27 에서 다룬다* 로 함께 정리합니다. 다만 이 경우 챕터의 "규모 격차" 비교가 사라지므로 1안을 권합니다.
3. 어느 쪽이든 **WikiDocs 인정 문단의 틀린 원인 설명은 제거**해야 합니다 (산문 수정 범위 안).

## 재현

```bash
git checkout d6c78e1
FORCE=1 BRANCH=master ./executed/run_via_cli.sh 26     # colab-cli, T4
# executed/26_ko_tiny_gpt.ipynb 의 c39 출력 확인
```

**전파 경로**: 노트북 `c39` → `executed/` → `pages/26-ko_tiny_gpt-variation.md` (라이브: https://wikidocs.net/370143 ) → `book/chapters/ch26_ko_tiny_gpt.tex` → EPUB `ch032.xhtml`. **세 산출물 모두 오염.**

**심각도**: Major. §8 이 "(선택)" 셀이고 챕터의 핵심 결론(scratch 사전학습 전·후 비교)은 영향받지 않아 Blocker 로 올리지는 않았습니다. 다만 *깨진 출력 + 정반대 서술 + 틀린 원인 진단* 이 세 산출물에 함께 실려 있어, 재검수자가 Blocker 로 판단하면 이견 없습니다.

Related to #50 (챕터 이슈). 참고: Ch 27 도 같은 `skt/kogpt2-base-v2` 를 쓰지만 커밋 실행본(2026-06-22)은 정상이라 현재 해당 없음 — 다만 Ch 27 을 transformers 5.15 로 재실행할 때 같은 증상이 나오는지 확인이 필요합니다 (D 페어 공유용).

**Comment by yum-yeom** [comment]
**원인 확정 — 토크나이저 fallback (본문의 safetensors 단서는 무관)**

transformers 5.15.0 (Colab 과 동일) 을 로컬에 두고 재현했습니다.
- `AutoTokenizer.from_pretrained("skt/kogpt2-base-v2")` → **`GPT2Tokenizer`(영어, vocab 51,201, eos `<|endoftext|>`)** 로 로드됨. `"옛날 옛날에"` → ids `[501, 500, 529, …]` → decode `'�����'` — prompt 인코딩 단계에서 이미 깨짐.
- 가중치는 정상 적재 (LOAD REPORT 의 UNEXPECTED 는 `masked_bias` 뿐, `wte` std 0.0375). 본문에서 "가중치 미적재" 로 의심한 safetensors 재다운로드는 원인이 아닙니다.
- Ch 27 이 이미 같은 이유로 `PreTrainedTokenizerFast.from_pretrained(..., bos_token="</s>", eos_token="</s>", unk_token="<unk>", pad_token="<pad>", mask_token="<mask>")` 를 쓰고 있고(주석에 "AutoTokenizer 가 영어 GPT2 로 fallback" 명시), 같은 패턴을 5.15 에 적용하면 `[12346, 35970]` → `'옛날 옛날에'` round-trip + 정상 한국어 생성 확인 (`옛날 옛날에 이게 네. 그게 막 그~ …`, Ch 27 실행본과 같은 결).

→ **제안 1안으로 수정**: `c39` 를 Ch 27 과 같은 로드 패턴으로 교체 (`e45a1e8`, 브랜치 `ch-26/review-2026-09`). Colab T4 재실행 후 실행본·WikiDocs(인정 문단의 틀린 원인 설명 제거)·EPUB 갱신은 PR 로 이어집니다. CLAUDE.md 의 "토크나이저 fallback (Ch 27, Blocker)" 패턴과 동일 계열입니다.
